tests/helpers.py

```diff
@@ -203,3 +206,32 @@ def verify_reboot_performed(self, max_wait=60*30):
     def verify_reboot_not_performed(self, wait=60):
         time.sleep(wait)
         assert exists(self.tfile)
+
+    def verify_device_double_reboot(self, wait=60*5):
+        self.dbootfile = "/mender-testing.%s" % (random.randint(1, 999999))
+        cmd = "touch %s" % (self.dbootfile)
+        with settings(hide('warnings', 'running', 'stdout', 'stderr'), abort_exception=FabricFatalException):
+            run(cmd)
+        # no time to wait for successfull connections
+        # mender never enters state-machine on the new partition
+        self.verify_reboot_performed(ntimes=0, sleeptime=3)
+        logger.debug("reboot in progress")
+        timeout = time.time() + wait
+        # wait until the dbootfile reappears
+        while time.time() <= timeout:
+            try:
+                with settings(warn_only=True, abort_exception=FabricFatalException):
+                    time.sleep(3)
+                    if not exists(self.dbootfile):
+                        logger.debug("reboot file has not reappeared, waiting")
+                        continue
+                    else:
+                        logger.debug("reboot file has reappeared. success!")
+                        return
+            except (BaseException):
+                logger.debug("no SSH connection. reboot in progress...")
+                continue
```

Comment:
In general a double reboot detector is going to have race conditions because the agent might be reboot before we have time to check whether it rebooted. So this needs to be reworked somehow. Good news is that I'm deep into reboot semantics right now, so I can probably help you with this one. I'm also fairly certain that this is not the only place that tries to detect double reboot.